### PR TITLE
Add 'WithDictIndex' expiry API and update RANDOMKEY command

### LINKED ISSUE #1145: [CRASH] Cluster Mode Cross-Slot RANDOMKEY in MULTI-EXEC
### Crash report

```shell
Logged crash report (pid 23580):
=== VALKEY BUG REPORT START: Cut & paste starting from here ===
23580:M 10 Oct 2024 19:36:19.133 # === ASSERTION FAILED CLIENT CONTEXT ===
23580:M 10 Oct 2024 19:36:19.133 # client->flags = 18023196754182152
23580:M 10 Oct 2024 19:36:19.133 # client->conn = fd=12
23580:M 10 Oct 2024 19:36:19.133 # client->argc = 1
23580:M 10 Oct 2024 19:36:19.133 # client->argv[0] = "RANDOMKEY" (refcount: 1)
23580:M 10 Oct 2024 19:36:19.133 # === RECURSIVE ASSERTION FAILED ===
23580:M 10 Oct 2024 19:36:19.133 # ==> db.c:246 '(int)keyHashSlot(key, (int)sdslen(key)) == server.current_client->slot' is not true

------ STACK TRACE ------

23587 bio_lazy_free
/lib64/libpthread.so.0(pthread_cond_wait+0x21c)[0xffff9c212e3c]
src/valkey-server 127.0.0.1:21116 [cluster](bioProcessBackgroundJobs+0x144)[0x4cc0f4]
/lib64/libpthread.so.0(+0x7230)[0xffff9c20c230]
/lib64/libc.so.6(+0xdb7dc)[0xffff9c15a7dc]

23586 bio_aof
/lib64/libpthread.so.0(pthread_cond_wait+0x21c)[0xffff9c212e3c]
src/valkey-server 127.0.0.1:21116 [cluster](bioProcessBackgroundJobs+0x144)[0x4cc0f4]
/lib64/libpthread.so.0(+0x7230)[0xffff9c20c230]
/lib64/libc.so.6(+0xdb7dc)[0xffff9c15a7dc]

23585 bio_close_file
/lib64/libpthread.so.0(pthread_cond_wait+0x21c)[0xffff9c212e3c]
src/valkey-server 127.0.0.1:21116 [cluster](bioProcessBackgroundJobs+0x144)[0x4cc0f4]
/lib64/libpthread.so.0(+0x7230)[0xffff9c20c230]
/lib64/libc.so.6(+0xdb7dc)[0xffff9c15a7dc]

23580 valkey-server *
src/valkey-server 127.0.0.1:21116 [cluster](getKeySlot+0x214)[0x469654]
src/valkey-server 127.0.0.1:21116 [cluster][0x46968c]
src/valkey-server 127.0.0.1:21116 [cluster](dbRandomKey+0xac)[0x470db8]
src/valkey-server 127.0.0.1:21116 [cluster](randomkeyCommand+0x18)[0x470e7c]
src/valkey-server 127.0.0.1:21116 [cluster](call+0x760)[0x5461c0]
src/valkey-server 127.0.0.1:21116 [cluster](execCommand+0x268)[0x45d4f0]
src/valkey-server 127.0.0.1:21116 [cluster](call+0x760)[0x5461c0]
src/valkey-server 127.0.0.1:21116 [cluster](processCommand+0xb7c)[0x497d1c]
src/valkey-server 127.0.0.1:21116 [cluster](processInputBuffer+0x180)[0x4b7668]
src/valkey-server 127.0.0.1:21116 [cluster](readQueryFromClient+0x50)[0x4b9c50]
src/valkey-server 127.0.0.1:21116 [cluster][0x53c6b8]
src/valkey-server 127.0.0.1:21116 [cluster](aeProcessEvents+0xdc)[0x574d80]
src/valkey-server 127.0.0.1:21116 [cluster](aeMain+0x28)[0x5753ac]
src/valkey-server 127.0.0.1:21116 [cluster](main+0x3ac)[0x45090c]
/lib64/libc.so.6(__libc_start_main+0xe4)[0xffff9c09eda4]
src/valkey-server 127.0.0.1:21116 [cluster][0x450ff0]
```
```


### Additional information

```shell
Fails the following test

start_cluster 1 0 {tags {external:skip cluster}} {
    test "Regression test for multi-exec with RANDOMKEY accessing the wrong per-slot dicitonary" {
        R 0 SETEX FOO 10000 BAR
        R 0 SETEX FIZZ 10000 BUZZ

        R 0 MULTI
        R 0 GET FOO
        R 0 RANDOMKEY
        set result [R 0 EXEC]
        assert_equal $result {{} FOO}
    }
}
```
```

### PR DESCRIPTION
**Pull request by nadav-levanoni**
https://github.com/valkey-io/valkey/issues/1145

First part of a two-step effort to add `WithSlot` API for expiry. This PR is to fix a crash that occurs when a RANDOMKEY uses a different slot than the cached slot of a client during a multi-exec.

The next part will be to utilize the new API as an optimization to prevent duplicate work when calculating the slot for a key.

### CODE REVIEW

File: tests/unit/cluster/multi-slot-operations.tcl [hunk 1]
```diff
@@ -107,3 +107,15 @@ test "DELSLOTSRANGE command with several boundary conditions test suite" {
     assert_match "*9829 11000*12001 12100*12201 13104*" [$master4 CLUSTER SLOTS]
 }
 } cluster_allocate_with_continuous_slots_local
+
+start_cluster 1 0 {tags {external:skip cluster}} {
+    test "Regression test for multi-exec with RANDOMKEY accessing the wrong per-slot dicitonary" {
```
Comment (madolson) [inline tests/unit/cluster/multi-slot-operations.tcl]:
```suggestion
    test "Regression test for multi-exec with RANDOMKEY accessing the wrong per-slot dictionary" {
```

File: src/server.h [hunk 2]
```diff
@@ -3338,6 +3338,7 @@ int calculateKeySlot(sds key);
 int dbExpand(serverDb *db, uint64_t db_size, int try_expand);
 int dbExpandExpires(serverDb *db, uint64_t db_size, int try_expand);
 dictEntry *dbFind(serverDb *db, void *key);
+dictEntry *dbFindExpiresWithSlot(serverDb *db, void *key, int slot);
```
Comment (madolson) [inline src/server.h]: This probably shouldn't be exposed, we have no other APIs that expose the slot.

File: tests/unit/cluster/multi-slot-operations.tcl [hunk 3]
```diff
@@ -107,3 +107,15 @@ test "DELSLOTSRANGE command with several boundary conditions test suite" {
...
+
+start_cluster 1 0 {tags {external:skip cluster}} {
+    test "Regression test for multi-exec with RANDOMKEY accessing the wrong per-slot dictionary" {
+        R 0 SETEX FOO 10000 BAR
+        R 0 SETEX FIZZ 10000 BUZZ
+
+        R 0 MULTI
+        R 0 GET FOO
+        R 0 RANDOMKEY
+        R 0 EXEC
```
Comment (madolson) [inline tests/unit/cluster/multi-slot-operations.tcl]: I think this test won't deterministically fail, if the randomkey hits FOO, then the slot will be the same. 
```suggestion
        R 0 SETEX FIZZ 10000 BUZZ

        R 0 MULTI
        R 0 GET FOO
        R 0 RANDOMKEY
        assert_equal [R 0 EXEC] {{} FIZZ}
```
I also think it would still be better to assert the response. I think this is right.

File: src/db.c [hunk 4]
```diff
@@ -1693,14 +1695,21 @@ void setExpire(client *c, serverDb *db, robj *key, long long when) {
...
-    if ((de = dbFindExpires(db, key->ptr)) == NULL) return -1;
+    if ((de = dbFindExpiresWithSlot(db, key->ptr, slot)) == NULL) return -1;
 
     return dictGetSignedIntegerVal(de);
 }
 
+/* Return the expire time of the specified key, or -1 if no expire
+ * is associated with this key (i.e. the key is non volatile) */
+long long getExpire(serverDb *db, robj *key) {
+    int slot = server.cluster_enabled ? getKeySlot(key->ptr) : 0;
```
Comment (madolson) [inline src/db.c]:
```suggestion
    int dict_index = getKVStoreIndexForKey(key->ptr);
```
Throughout the codebase I think the preference was to call this KVStore index as opposed to slot. It is the slot for cluster mode, but it's not the slot in CMD.

**Review by madolson — COMMENTED** [review]
The code looks OK to me, I made a general comment about using dict_index and kvstore index, since is a bit more consistent within the file.

**Comment by zuiderkwast** [comment]
@nadav-levanoni Please fix the DCO issue, as described in the Details link next to the failed DCO CI job (or [here](https://github.com/valkey-io/valkey/pull/1155/checks?check_run_id=31432668604)). What is DCO and why we need it is described in CONTRIBUTING.md.

File: tests/unit/cluster/multi-slot-operations.tcl (hunk 3 above)
Comment (zuiderkwast) [reply tests/unit/cluster/multi-slot-operations.tcl]: This test suite is about admin commands like CLUSTER ADDSLOTRANGE, etc. I don't think it was intended for things like RANDOMKEY in MULTI.

RANDOMKEY is tested in unit/keyspace.tcl. MULTI is tested in unit/multi.tcl. One of those would be a better place for this test case IMO.

Comment (nadav-levanoni) [reply tests/unit/cluster/multi-slot-operations.tcl]: The issue is specific to cluster mode enabled. Do we have the capability to enable it in unit/keyspace.tcl?

Comment (zuiderkwast) [reply tests/unit/cluster/multi-slot-operations.tcl]: Yea i believe it's possible. Just add a tag I think. I'll check tomorrow.

Comment (nadav-levanoni) [reply tests/unit/cluster/multi-slot-operations.tcl]: I can play around with the test. Maybe something like this would be deterministic
```
        R 0 SETEX FOO 10000 BAR
        R 0 SETEX FIZZ 10000 BUZZ
        R 0 MULTI
        R 0 DEL FOO
        R 0 RANDOMKEY
        assert_equal [R 0 EXEC] {1 FIZZ}
```

The cached slot will be of FOO and then RANDOMKEY must select FIZZ.

Comment (zuiderkwast) [reply tests/unit/cluster/multi-slot-operations.tcl]: I see test cases under files like `unit/other.tcl` use this style to run in cluster mode. (Actually exactly as you did here.) I think `unit/multi.tcl` is the  best place.

```tcl
start_cluster 1 0 {tags {"external:skip cluster"}} {
    test "Bla bla bla" {
        r flushall
        # ...        
    }
}
```

Comment (madolson) [reply tests/unit/cluster/multi-slot-operations.tcl]: It's only cluster specific, so wanted it to be in the cluster subdirectory. I originally thought it maybe should be in cluster/misc. I don't feel strongly though.

Comment (nadav-levanoni) [reply tests/unit/cluster/multi-slot-operations.tcl]: Sweet, moved it to multi.tcl and used your configuration and it seems like it produces a consistent crash

File: src/db.c (hunk 4 above)
Comment (nadav-levanoni) [reply src/db.c]: I'll update the usage

File: src/db.c [hunk 5]
```diff
@@ -375,13 +378,13 @@ robj *dbRandomKey(serverDb *db) {
     while (1) {
         sds key;
         robj *keyobj;
-        int randomSlot = kvstoreGetFairRandomDictIndex(db->keys);
-        de = kvstoreDictGetFairRandomKey(db->keys, randomSlot);
+        int randomDbIndex = kvstoreGetFairRandomDictIndex(db->keys);
+        de = kvstoreDictGetFairRandomKey(db->keys, randomDbIndex);
```
Comment (madolson) [inline src/db.c]:
```suggestion
        int randomDictIndex = kvstoreGetFairRandomDictIndex(db->keys);
        de = kvstoreDictGetFairRandomKey(db->keys, randomDictIndex);
```

File: src/db.c [hunk 6]
```diff
@@ -375,13 +378,13 @@ robj *dbRandomKey(serverDb *db) {
...
         if (de == NULL) return NULL;
 
         key = dictGetKey(de);
         keyobj = createStringObject(key, sdslen(key));
-        if (dbFindExpires(db, key)) {
+        if (dbFindExpiresWithDictIndex(db, key, randomDbIndex)) {
```
Comment (madolson) [inline src/db.c]:
```suggestion
        if (dbFindExpiresWithDictIndex(db, key, randomDictIndex)) {
```

File: src/db.c [hunk 7]
```diff
@@ -393,7 +396,7 @@ robj *dbRandomKey(serverDb *db) {
                  * return a key name that may be already expired. */
                 return keyobj;
             }
-            if (expireIfNeeded(db, keyobj, 0) != KEY_VALID) {
+            if (expireIfNeededWithDictIndex(db, keyobj, 0, randomDbIndex) != KEY_VALID) {
```
Comment (madolson) [inline src/db.c]:
```suggestion
            if (expireIfNeededWithDictIndex(db, keyobj, 0, randomDictIndex) != KEY_VALID) {
```

**Review by madolson — APPROVED** [review]
Thanks!
